### ISSUE TITLE
Error loading workspace: packages.Load error: err: exit status 2: stderr: panic: runtime error: index out of range [-1]

### ISSUE THREAD

**Issue by vasily-kirichenko** [golang/vscode-go#2956]
### What version of Go, VS Code & VS Code Go extension are you using?
<details><summary>Version Information</summary><br>

* Run `go version` to get version of Go from _the VS Code integrated terminal_.
	- go version go1.21.0 linux/amd64
* Run `gopls -v version` to get version of Gopls from _the VS Code integrated terminal_.
	- Build info
----------
golang.org/x/tools/gopls v0.13.2
    golang.org/x/tools/gopls@v0.13.2 h1:Pyvx6MKvatbX3zzZmdGiFRfQZl0ohPlt2sFxO/5j6Ro=
    github.com/BurntSushi/toml@v1.2.1 h1:9F2/+DoOYIOksmaJFPw1tGFy1eDnIJXg+UHjuD8lTak=
    github.com/google/go-cmp@v0.5.9 h1:O2Tfq5qg4qc4AmwVlvv0oLiVAGB7enBSJ2x2DqQFi38=
    github.com/sergi/go-diff@v1.1.0 h1:we8PVUC3FE2uYfodKH/nBHMSetSfHDR6scGdBi+erh0=
    golang.org/x/exp@v0.0.0-20220722155223-a9213eeb770e h1:+WEEuIdZHnUeJJmEUjyYC2gfUMj69yZXw17EnHg/otA=
    golang.org/x/exp/typeparams@v0.0.0-20221212164502-fae10dda9338 h1:2O2DON6y3XMJiQRAS1UWU+54aec2uopH3x7MAiqGW6Y=
    golang.org/x/mod@v0.12.0 h1:rmsUpXtvNzj340zd98LZ4KntptpfRHwpFOHG188oHXc=
    golang.org/x/sync@v0.3.0 h1:ftCYgMx6zT/asHUrPw8BLLscYtGznsLAnjq5RH9P66E=
    golang.org/x/sys@v0.10.0 h1:SqMFp9UcQJZa+pmYuAKjd9xq1f0j5rLcDIk0mj4qAsA=
    golang.org/x/text@v0.11.0 h1:LAntKIrcmeSKERyiOh0XMV39LXS8IE9UL2yP7+f5ij4=
    golang.org/x/tools@v0.11.2-0.20230810185051-cc6b5804b8cf h1:Oush7UwPamr2/iNeNFBuNFj89YyHn0YY69EKDdvANnk=
    golang.org/x/vuln@v0.0.0-20230110180137-6ad3e3d07815 h1:A9kONVi4+AnuOr1dopsibH6hLi1Huy54cbeJxnq4vmU=
    honnef.co/go/tools@v0.4.2 h1:6qXr+R5w+ktL5UkwEbPp+fEvfyoMPche6GkOpGHZcLc=
    mvdan.cc/gofumpt@v0.4.0 h1:JVf4NN1mIpHogBj7ABpgOyZc65/UUOkKQFkoURsz4MM=
    mvdan.cc/xurls/v2@v2.4.0 h1:tzxjVAj+wSBmDcF6zBB7/myTy3gX9xvi8Tyr28AuQgc=
go: go1.21.0
* Run `code -v` or `code-insiders -v` to get version of VS Code or VS Code Insiders.
	- 1.81.1
6c3e3dba23e8fadc360aed75ce363ba185c49794
x64

code-insiders: command not found

* Check your installed extensions to get the version of the VS Code Go extension
	- v0.39.1
* Run Ctrl+Shift+P (Cmd+Shift+P on Mac OS) > `Go: Locate Configured Go Tools` command.
	- Checking configured tools....
GOBIN: undefined
toolsGopath: 
gopath: /home/kirichenko/go
GOROOT: /usr/local/go
PATH: /home/kirichenko/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/home/kirichenko/.local/share/JetBrains/Toolbox/scripts:/usr/local/go/bin:/home/kirichenko/dotnet:/home/kirichenko/.local/protoc/bin:/home/kirichenko/.local/share/JetBrains/Toolbox/scripts:/usr/local/go/bin:/home/kirichenko/dotnet:/home/kirichenko/go/bin:/home/kirichenko/.dotnet/tools:/home/kirichenko/.local/protoc/bin:/home/kirichenko/.local/share/coursier/bin:/opt/oc/openshift-origin-client-tools-v3.11.0-0cbc58b-linux-64bit

	go:	/usr/local/go/bin/go: go version go1.21.0 linux/amd64

	gotests:	/home/kirichenko/go/bin/gotests	(version: v1.6.0 built with go: go1.21.0)
	gomodifytags:	/home/kirichenko/go/bin/gomodifytags	(version: v1.16.0 built with go: go1.21.0)
	impl:	/home/kirichenko/go/bin/impl	(version: v1.1.0 built with go: go1.21.0)
	goplay:	/home/kirichenko/go/bin/goplay	(version: v1.0.0 built with go: go1.21.0)
	dlv:	/home/kirichenko/go/bin/dlv	(version: v1.21.0 built with go: go1.21.0)
	golangci-lint:	/home/kirichenko/go/bin/golangci-lint	(version: v1.54.0 built with go: go1.21.0)
	gopls:	/home/kirichenko/go/bin/gopls	(version: v0.13.2 built with go: go1.21.0)

go env
Workspace Folder (go): /home/kirichenko/f/go
	GO111MODULE=''
	GOARCH='amd64'
	GOBIN=''
	GOCACHE='/home/kirichenko/.cache/go-build'
	GOENV='/home/kirichenko/.config/go/env'
	GOEXE=''
	GOEXPERIMENT=''
	GOFLAGS=''
	GOHOSTARCH='amd64'
	GOHOSTOS='linux'
	GOINSECURE=''
	GOMODCACHE='/home/kirichenko/go/pkg/mod'
	GONOPROXY='none'
	GONOSUMDB='go.kl/*'
	GOOS='linux'
	GOPATH='/home/kirichenko/go'
	GOPRIVATE='go.kl/*'
	GOPROXY='https://<private url here>' - I edited this line manually to hide private info
	GOROOT='/usr/local/go'
	GOSUMDB='sum.golang.org'
	GOTMPDIR=''
	GOTOOLCHAIN='auto'
	GOTOOLDIR='/usr/local/go/pkg/tool/linux_amd64'
	GOVCS=''
	GOVERSION='go1.21.0'
	GCCGO='gccgo'
	GOAMD64='v1'
	AR='ar'
	CC='gcc'
	CXX='g++'
	CGO_ENABLED='1'
	GOMOD='/home/kirichenko/f/go/go.mod'
	GOWORK=''
	CGO_CFLAGS='-O2 -g'
	CGO_CPPFLAGS=''
	CGO_CXXFLAGS='-O2 -g'
	CGO_FFLAGS='-O2 -g'
	CGO_LDFLAGS='-O2 -g'
	PKG_CONFIG='pkg-config'
	GOGCCFLAGS='-fPIC -m64 -pthread -Wl,--no-gc-sections -fmessage-length=0 -ffile-prefix-map=/tmp/go-build2160459470=/tmp/go-build -gno-record-gcc-switches'
</details>

### Share the Go related settings you have added/edited

Run `Preferences: Open Settings (JSON)` command to open your settings.json file.
Share all the settings with the `go.` or `["go"]` or `gopls` prefixes.

    "gopls": {
        "ui.codelenses": {
            "gc_details": false,
            "upgrade_dependency": true
        },
        "ui.diagnostic.annotations": {
            "escape": true,
            "bounds": true,
            "inline": true,
            "nil": true
        },
        "ui.diagnostic.analyses": {
            "nilness": true,
            "unusedparams": true,
            "unusedvariable": true,
            "unusedwrite": true
        },
        "ui.diagnostic.staticcheck": true,
        "formatting.gofumpt": true,
        "ui.completion.usePlaceholders": false,
        "ui.completion.completionBudget": "100ms"
    },
    "go.inlayHints.assignVariableTypes": true,
    "go.inlayHints.rangeVariableTypes": true,
    "go.inlayHints.parameterNames": true,
    "go.inlayHints.functionTypeParameters": true,
    "go.inlayHints.compositeLiteralFields": true,
    "go.inlayHints.compositeLiteralTypes": true,
    "go.inlayHints.constantValues": true,
    "go.toolsManagement.autoUpdate": true,
    "go.lintTool": "golangci-lint",
    "go.lintFlags": [
        "--fast"
        //"--disable=typecheck"
    ],
    "go.lintOnSave": "package",
    "go.testExplorer.showDynamicSubtestsInEditor": true,
    "go.testFlags": [
        "-v",
        "-count=1",
        "-timeout=0"
    ],
    "go.addTags": {
        "tags": "json,yaml",
        "promptForTags": true,
    },
    "go.languageServerExperimentalFeatures": {
        "diagnostics": true
    },
    "go.editorContextMenuCommands": {
        "toggleTestFile": true,
        "addTags": true,
        "removeTags": true,
        "fillStruct": false,
        "testAtCursor": true,
        "testFile": true,
        "testPackage": false,
        "generateTestForFunction": true,
        "generateTestForFile": false,
        "generateTestForPackage": false,
        "addImport": true,
        "testCoverage": true,
        "playground": true,
        "debugTestAtCursor": true,
        "benchmarkAtCursor": false
    },
    "[go]": {
        "editor.formatOnSave": true
    },
    "go.disableConcurrentTests": true,
    "go.playground": {
        "share": false
    },

### Describe the bug

No gopls features works, the following error is infinitely spinning in the status bar:

```
Error loading workspace: packages.Load error: err: exit status 2: stderr: panic: runtime error: index out of range [-1]  goroutine 1 [running]: cmd/go/internal/list.collectDepsErrors.func1(0xc00a179968?, 0xc006b0eff0?) 	cmd/go/internal/list/list.go:964 +0x134 sort.insertionSort_func({0xc006b0f110?, 0xc004307100?}, 0x0, 0x3) 	sort/zsortfunc.go:12 +0xa7 sort.pdqsort_func({0xc006b0f110?, 0xc004307100?}, 0x7fc7e748ecc8?, 0x18?, 0xc000eda000?) 	sort/zsortfunc.go:73 +0x31b sort.Slice({0x9acea0?, 0xc00a179968?}, 0x3?) 	sort/slice.go:26 +0xf6 cmd/go/internal/list.collectDepsErrors(0xc002db9800) 	cmd/go/internal/list/list.go:953 +0x2fe cmd/go/internal/list.runList({0xb85310?, 0xf21ee0}, 0xee5920?, {0xc0000221c0?, 0x2, 0x2}) 	cmd/go/internal/list/list.go:809 +0x21a6 main.invoke(0xee5920, {0xc000022100, 0xe, 0xe}) 	cmd/go/main.go:268 +0x5f1 main.main() 	cmd/go/main.go:186 +0x7a5 
```

### Steps to reproduce the behavior:
1. Open the certain directory with a go module or execute "restart language server" action. The bug is not reproduced on a hello world module and some simpler modules. 

```
[Info  - 2:11:04 PM] 2023/08/29 14:11:04 go info for /home/kirichenko/f/go
(go dir /home/kirichenko/f/go)
(go version go version go1.21.0 linux/amd64)
(valid build configuration = true)
(build flags: [])
(selected go env: [GO111MODULE=, GOCACHE=/home/kirichenko/.cache/go-build, GOFLAGS=, GOMODCACHE=/home/kirichenko/go/pkg/mod, GOPATH=/home/kirichenko/go, GOPRIVATE=go.kl/*, GOROOT=/usr/local/go, GOWORK=])


[Error - 2:11:06 PM] 2023/08/29 14:11:06 go/packages.Load #1: err: exit status 2: stderr: panic: runtime error: index out of range [-1]

goroutine 1 [running]:
cmd/go/internal/list.collectDepsErrors.func1(0xc0082461b0?, 0xc004d06ff0?)
	cmd/go/internal/list/list.go:964 +0x134
sort.insertionSort_func({0xc004d07110?, 0xc003fc4ba0?}, 0x0, 0x3)
	sort/zsortfunc.go:12 +0xa7
sort.pdqsort_func({0xc004d07110?, 0xc003fc4ba0?}, 0x7f5fe782c688?, 0x18?, 0xc000600000?)
	sort/zsortfunc.go:73 +0x31b
sort.Slice({0x9acea0?, 0xc0082461b0?}, 0x3?)
	sort/slice.go:26 +0xf6
cmd/go/internal/list.collectDepsErrors(0xc005238c00)
	cmd/go/internal/list/list.go:953 +0x2fe
cmd/go/internal/list.runList({0xb85310?, 0xf21ee0}, 0xee5920?, {0xc0000221c0?, 0x2, 0x2})
	cmd/go/internal/list/list.go:809 +0x21a6
main.invoke(0xee5920, {0xc000022100, 0xe, 0xe})
	cmd/go/main.go:268 +0x5f1
main.main()
	cmd/go/main.go:186 +0x7a5

	snapshot=0
	directory=file:///home/kirichenko/f/go
	query=[/home/kirichenko/f/go/... builtin]
	packages=0

[Error - 2:11:06 PM] 2023/08/29 14:11:06 initial workspace load failed: packages.Load error: err: exit status 2: stderr: panic: runtime error: index out of range [-1]

goroutine 1 [running]:
cmd/go/internal/list.collectDepsErrors.func1(0xc0082461b0?, 0xc004d06ff0?)
	cmd/go/internal/list/list.go:964 +0x134
sort.insertionSort_func({0xc004d07110?, 0xc003fc4ba0?}, 0x0, 0x3)
	sort/zsortfunc.go:12 +0xa7
sort.pdqsort_func({0xc004d07110?, 0xc003fc4ba0?}, 0x7f5fe782c688?, 0x18?, 0xc000600000?)
	sort/zsortfunc.go:73 +0x31b
sort.Slice({0x9acea0?, 0xc0082461b0?}, 0x3?)
	sort/slice.go:26 +0xf6
cmd/go/internal/list.collectDepsErrors(0xc005238c00)
	cmd/go/internal/list/list.go:953 +0x2fe
cmd/go/internal/list.runList({0xb85310?, 0xf21ee0}, 0xee5920?, {0xc0000221c0?, 0x2, 0x2})
	cmd/go/internal/list/list.go:809 +0x21a6
main.invoke(0xee5920, {0xc000022100, 0xe, 0xe})
	cmd/go/main.go:268 +0x5f1
main.main()
	cmd/go/main.go:186 +0x7a5


[Error - 2:11:06 PM] 2023/08/29 14:11:06 errors loading workspace: packages.Load error: err: exit status 2: stderr: panic: runtime error: index out of range [-1]

goroutine 1 [running]:
cmd/go/internal/list.collectDepsErrors.func1(0xc0082461b0?, 0xc004d06ff0?)
	cmd/go/internal/list/list.go:964 +0x134
sort.insertionSort_func({0xc004d07110?, 0xc003fc4ba0?}, 0x0, 0x3)
	sort/zsortfunc.go:12 +0xa7
sort.pdqsort_func({0xc004d07110?, 0xc003fc4ba0?}, 0x7f5fe782c688?, 0x18?, 0xc000600000?)
	sort/zsortfunc.go:73 +0x31b
sort.Slice({0x9acea0?, 0xc0082461b0?}, 0x3?)
	sort/slice.go:26 +0xf6
cmd/go/internal/list.collectDepsErrors(0xc005238c00)
	cmd/go/internal/list/list.go:953 +0x2fe
cmd/go/internal/list.runList({0xb85310?, 0xf21ee0}, 0xee5920?, {0xc0000221c0?, 0x2, 0x2})
	cmd/go/internal/list/list.go:809 +0x21a6
main.invoke(0xee5920, {0xc000022100, 0xe, 0xe})
	cmd/go/main.go:268 +0x5f1
main.main()
	cmd/go/main.go:186 +0x7a5

	snapshot=0
	directory=file:///home/kirichenko/f/go 
```

**Comment by hyangah** [comment]
@vasily-kirichenko Can you try `go list  /home/kirichenko/f/go/... builtin` from /home/kirichenko/f/go directory?

cc Looks like a bug in `go list` code @bcmills @matloob

**Comment by vasily-kirichenko** [comment]
@hyangah yes, go list fails the same way. I compiled go from master, it works. After that I pointed the extension to that fresh go, the results were strange: although there were no errors at all, nothing works on some packages (dirs), even symbol under cursor was not highlighted.

**Comment by findleyr** [comment]
Thanks for the report. This is the Go command bug https://go.dev/issue/61816, which is fixed at tip. The fix should be released in Go 1.21.1, which I believe is scheduled to be released within the next week or two.
